### ISSUE TITLE
Slower due to Ipad iOS 17.4

### ISSUE THREAD

**Issue by BeezBumba** [nbarkhina/N64Wasm#43]
When I updated my Ipad to 17.4 I went to my remixed emulator of yours and it started to lag much more than last time. I even tried yours, but it still resulted into a really laggy experience. I don’t know if it’s the update or if it’s something else.

**Comment by nbarkhina** [comment]
Were you able to fix this? I didn't test on 17.4 yet

**Comment by BeezBumba** [comment]
No, but it might need tweaking on the webkit feature flags for safari I believe. I’m not sure but it might be a way.

**Comment by nbarkhina** [comment]
Yeah I just tried it this morning on 17.4 on my iPhone 13 mini and can confirm it is really laggy now. Seems like Apple changed something in Safari. I will also play around with the feature flags and see if there is any workaround for this.

**Comment by BeezBumba** [comment]
It might be a CPU problem in there. I know they added WebGPU, WebRTC stuff, and 2 or more WebGL settings. It was not talked about in the update and it’s probably something directly on the webkit website.

**Comment by BeezBumba** [comment]
Did you find anything that may help yet?

**Comment by nbarkhina** [comment]
Nope nothing yet

**Comment by BeezBumba** [comment]
I have sent a website issue note to apple because now I'm not really sure what could be the issue. I am suspecting its memory only because when I was editing a Google Slides presentation, whenever I would copy and paste, it can take a while for it to paste.

**Comment by nbarkhina** [comment]
Great - let me know if they come back with anything. Yeah my guess is there's not much we can do unless Apple fixes the issue.

**Comment by BeezBumba** [comment]
I now know that it’s not something with the feature flags. I tried enabling/disabling all of them to see if there is a difference, but there was no difference. I did have to leave some enabled so that it will work.

**Comment by BeezBumba** [comment]
One more thing. My app store is not working for some reason, so can you try to see if the website works for other browsers?

**Comment by nbarkhina** [comment]
I tried in Chrome but it's the same result. All iOS browsers use the same engine.

**Comment by BeezBumba** [comment]
People are saying that update 17.4.1 is supposed to come in a few days. Maybe, in that update the issue will be fixed?

**Comment by BeezBumba** [comment]
Quick question: Is there any part of your scripts that tries to sync the gameplay with the audio if the audio is acting slow? I saw a part in the script.js script that is highlighted and says that it’s outdated and use a different one.

**Comment by BeezBumba** [comment]
It also has benefits like working on a separate thread instead of the main thread.

**Comment by BeezBumba** [comment]
17.4.1 has just been released. Have you tried it yet to see if it fixed the issue?

**Comment by BeezBumba** [comment]
Also, I did find something intresting. On this website running Banjo-Kazooie, It ran much smoother than this one. Here is the website: https://selenite.is-a.dev/semag/banjokazooie/index.html. I’m pretty sure gamesfrog.com has that engine too, but it was taking a little long for it to load so I’m not sure if it’s the same on gamesfrog.

**Comment by nbarkhina** [comment]
17.4.1 did not help

**Comment by BeezBumba** [comment]
What were the other emulates you’ve tried?

**Comment by nbarkhina** [comment]
the one you linked to. if you wait a minute after the intro screen it starts lagging

**Comment by nbarkhina** [comment]
point being I wouldn't call it "playable". perhaps some screens with less geometry run ok, but get to the outside and its very bad. this really needs to be fixed on Apple's side, I don't think any little optimizations here are sufficient.

**Comment by BeezBumba** [comment]
Did you find this blog: https://webkit.org/blog/15063/webkit-features-in-safari-17-4/ ? If it’s a problem that apple has to fix. It might be helpful to find the issue anyways.

**Comment by nbarkhina** [comment]
nothing in the article really sticks out to me unfortunately

**Comment by BeezBumba** [comment]
I just found this website https://demo.emulatorjs.org/ that ran fullspeed for me. It uses Retroarch as the emulator. This might have to do with the Webkit engine, Since that actually runs and renders the web pages. They could of changed something up with the power some libraries use, most likely the deprecated ones.

**Comment by nbarkhina** [comment]
Runs just as slow for me for some reason. I tried Mario 64 and only the save select screen runs fast, everything else slow.

**Comment by BeezBumba** [comment]
That’s weird. my Ipad 8th generation is running smoothly on it.

**Comment by BeezBumba** [comment]
Do you have any ideas still? Apple has not responded back yet and I'm completely oblivious of what could be the issue.

**Comment by nbarkhina** [comment]
Nope - does not seem to be much we can do unless Apple hopefully fixes this some day.

**Comment by nbarkhina** [comment]
@BeezBumba ok I may have a fix, stay tuned...

**Comment by nbarkhina** [comment]
ok I made some updates can you try this version and let me know how it runs for you?

https://neilb.net/n64wasm/vbo_test/

**Comment by BeezBumba** [comment]
It works! The only thing you have to do if request the mobile website by clicking the Aa on the left hand side on the address bar. But, thank you so much! What did you change?
![IMG_2423](https://github.com/nbarkhina/N64Wasm/assets/130922034/d740bf30-05e3-4a89-884b-b8a93d92a6a1)

**Comment by nbarkhina** [comment]
Take a look at my last commit - you can see the changes I had to make to fix it.

https://github.com/nbarkhina/N64Wasm/commit/aca7b9846cae3da1e1d46c5e63503c3ee983cf8f

For iOS devices I switched it to use something called "vertex buffer objects" which seems to resolve the rendering lag introduced in Safari.

**Comment by BeezBumba** [comment]
Awesome, I'm putting the changes in the codespace so it can generate the new js and wasm files right now.

**Comment by BeezBumba** [comment]
Can I also ask how you though of vertex buffer objects?

**Comment by nbarkhina** [comment]
I did some experimentation and isolated the slowness to the glDrawArrays call in geometry.c - that is where it draws all the triangles for the scene. So I was looking at ways to speed that up or make it more efficient. I remembered seeing the VBO code before but I wasn't sure what it did at the time so I commented it out. So I tried adding it back and to my surprise it worked. My understanding is that vertex buffer objects offer certain performance gains.

**Comment by BeezBumba** [comment]
What if i were to change the vertex buffer size. what would that do?

**Comment by nbarkhina** [comment]
Not sure - give it a shot and let me know

**Comment by nbarkhina** [comment]
I can't even remember if those were my todos - it sounds like I may have been experimenting with performance

**Comment by BeezBumba** [comment]
yeah those were the original page's todos

**Comment by BeezBumba** [comment]
Forgot to type in here that I’m trying to find another way to speed it up since on other browsers off Safari, it runs fullspeed. My guess is that some fix backfired and actually caused stuff like this to slowdown. I just scrolled down to the sections where it talked about JavaScript and WebAssembly, and WebGL. https://developer.apple.com/documentation/safari-release-notes/safari-17_4-release-notes

**Comment by BeezBumba** [comment]
Was it just the glDrawArrays that you found that was slow?

**Comment by nbarkhina** [comment]
I'm not sure what aspect specifically was slow. My iPhone is on 18.1 - I just tried it and it runs fine for me.

**Comment by BeezBumba** [comment]
It just that **most** games run full speed except for these ones that I’ve tried: NFL QB CLUB 2000 & 2001, and some sm64 mods. There might be more, but I’m pretty sure all of them have the same issue. Summing it up some games are still laggy and I’m wondering if there was anything else other than the glDrawArrays.

**Comment by BeezBumba** [comment]
It just that most games run full speed except for these ones that I’ve tried: NFL QB CLUB 2000 & 2001, and some sm64 mods. There might be more, but I’m pretty sure all of them have the same issue. Summing it up some games are still laggy and I’m wondering if there was anything else other than the glDrawArrays.

**Comment by BeezBumba** [comment]
After the iOS 18.2 update, I feel like there was a slight improvement to FPS for some games. I would say like a 20% increase.

**Comment by BeezBumba** [comment]
But I don’t know if it was that update or the iOS 18 versions before it.
